File: aos/util/trapezoid_profile.py
#!/usr/bin/python

import numpy

class TrapezoidProfile(object):
  """Computes a trapezoidal motion profile

  Attributes:
    _acceleration_time: the amount of time the robot will travel at the
        specified acceleration (s)
    _acceleration: the acceleration the robot will use to get to the target
        (unit/s^2)
    _constant_time: amount of time to travel at a constant velocity to reach
        target (s)
    _deceleration_time: amount of time to decelerate (at specified
        deceleration) to target (s)
    _deceleration: decceleration the robot needs to get to goal velocity
        (units/s^2)
    _maximum_acceleration: the maximum acceleration (units/s^2)
    _maximum_velocity: the maximum velocity (unit/s)
    _timestep: time between calls to Update (delta_time)
    _output: output array containing distance to goal and velocity
  """
  def __init__(self, delta_time):
    """Constructs a TrapezoidProfile.

    Args:
      delta_time: time between calls to Update (seconds)
    """
    self._acceleration_time = 0
    self._acceleration = 0
    self._constant_time = 0
    self._deceleration_time = 0
    self._deceleration = 0

    self._maximum_acceleration = 0
    self._maximum_velocity = 0
    self._timestep = delta_time

    self._output = numpy.array(numpy.zeros((2,1)))

  # Updates the state
  def Update(self, goal_position, goal_velocity):
    self._CalculateTimes(goal_position - self._output[0], goal_velocity)

    next_timestep = self._timestep

    # We now have the amount of time we need to accelerate to follow the
    # profile, the amount of time we need to move at constant velocity
    # to follow the profile, and the amount of time we need to decelerate to
    # follow the profile.  Do as much of that as we have time left in dt.
    if self._acceleration_time > next_timestep:
      self._UpdateVals(self._acceleration, next_timestep)
    else:
      self._UpdateVals(self._acceleration, self._acceleration_time)

      next_timestep -= self._acceleration_time
      if self._constant_time > next_timestep:
        self._UpdateVals(0, next_timestep)
      else:
        self._UpdateVals(0, self._constant_time)
        next_timestep -= self._constant_time;
        if self._deceleration_time > next_timestep:
          self._UpdateVals(self._deceleration, next_timestep)
        else:
          self._UpdateVals(self._deceleration, self._deceleration_time)
          next_timestep -= self._deceleration_time
          self._UpdateVals(0, next_timestep)

    return self._output

  # Useful for preventing windup etc.
  def MoveCurrentState(self, current):
    self._output = current

  # Useful for preventing windup etc.
  def MoveGoal(self, dx):
    self._output[0] += dx

  def SetGoal(self, x):
    self._output[0] = x

  def set_maximum_acceleration(self, maximum_acceleration):
    self._maximum_acceleration = maximum_acceleration

  def set_maximum_velocity(self, maximum_velocity):
    self._maximum_velocity = maximum_velocity

  def _UpdateVals(self, acceleration, delta_time):
    self._output[0, 0] += (self._output[1, 0] * delta_time
        + 0.5 * acceleration * delta_time * delta_time)
    self._output[1, 0] += acceleration * delta_time

  def _CalculateTimes(self, distance_to_target, goal_velocity):
    if distance_to_target == 0:
      self._acceleration_time = 0
      self._acceleration = 0
      self._constant_time = 0
      self._deceleration_time = 0
      self._deceleration = 0
      return
    elif distance_to_target < 0:
      # Recurse with everything inverted.
      self._output[1] *= -1
      self._CalculateTimes(-distance_to_target, -goal_velocity)
      self._output[1] *= -1
      self._acceleration *= -1
      self._deceleration *= -1
      return

    self._constant_time = 0
    self._acceleration = self._maximum_acceleration
    maximum_acceleration_velocity = (
        distance_to_target * 2 * numpy.abs(self._acceleration)
        + self._output[1] * self._output[1])
    if maximum_acceleration_velocity > 0:
      maximum_acceleration_velocity = numpy.sqrt(maximum_acceleration_velocity)
    else:
      maximum_acceleration_velocity = -numpy.sqrt(-maximum_acceleration_velocity)

    # Since we know what we'd have to do if we kept after it to decelerate, we
    # know the sign of the acceleration.
    if maximum_acceleration_velocity > goal_velocity:
      self._deceleration = -self._maximum_acceleration
    else:
      self._deceleration = self._maximum_acceleration

    # We now know the top velocity we can get to.
    top_velocity = numpy.sqrt((distance_to_target +
                                (self._output[1] * self._output[1]) /
                                (2.0 * self._acceleration) +
                                (goal_velocity * goal_velocity) /
                                (2.0 * self._deceleration)) /
                               (-1.0 / (2.0 * self._deceleration) +
                                1.0 / (2.0 * self._acceleration)))

    # If it can go too fast, we now know how long we get to accelerate for and
    # how long to go at constant velocity.
    if top_velocity > self._maximum_velocity:
      self._acceleration_time = ((self._maximum_velocity - self._output[1]) /
                                 self._maximum_acceleration)
      self._constant_time = (distance_to_target +
                        (goal_velocity * goal_velocity -
                         self._maximum_velocity * self._maximum_velocity) /
                        (2.0 * self._maximum_acceleration)) / self._maximum_velocity
    else:
      self._acceleration_time = (
          (top_velocity - self._output[1]) / self._acceleration)

    if self._output[1] > self._maximum_velocity:
      self._constant_time = 0
      self._acceleration_time = 0

    self._deceleration_time = (
        (goal_velocity - top_velocity) / self._deceleration)

File: aos/util/trapezoid_profile.cc
#include "aos/util/trapezoid_profile.h"

#include <math.h>

#include <algorithm>
#include <cstdlib>
#include <ostream>

#include "glog/logging.h"

#include "aos/time/time.h"

namespace aos::util {

AsymmetricTrapezoidProfile::AsymmetricTrapezoidProfile(
    ::std::chrono::nanoseconds delta_time)
    : timestep_(delta_time) {
  output_.setZero();
}

void AsymmetricTrapezoidProfile::UpdateVals(double acceleration,
                                            double delta_time) {
  output_(0) +=
      output_(1) * delta_time + 0.5 * acceleration * delta_time * delta_time;
  output_(1) += acceleration * delta_time;
}

const Eigen::Matrix<double, 2, 1> &AsymmetricTrapezoidProfile::Update(
    double goal_position, double goal_velocity) {
  CalculateTimes(goal_position - output_(0), goal_velocity, output_);

  double next_timestep = ::aos::time::DurationInSeconds(timestep_);

  if (deceleration_reversal_time_ > next_timestep) {
    UpdateVals(deceleration_reversal_, next_timestep);
    return output_;
  }

  UpdateVals(deceleration_reversal_, deceleration_reversal_time_);
  next_timestep -= deceleration_reversal_time_;

  if (acceleration_time_ > next_timestep) {
    UpdateVals(acceleration_, next_timestep);
    return output_;
  }

  UpdateVals(acceleration_, acceleration_time_);
  next_timestep -= acceleration_time_;

  if (constant_time_ > next_timestep) {
    UpdateVals(0, next_timestep);
    return output_;
  }

  UpdateVals(0, constant_time_);
  next_timestep -= constant_time_;
  if (deceleration_time_ > next_timestep) {
    UpdateVals(deceleration_, next_timestep);
  } else {
    UpdateVals(deceleration_, deceleration_time_);
    next_timestep -= deceleration_time_;
    UpdateVals(0, next_timestep);

    if (next_timestep >= 0 && goal_velocity == 0) {
      output_(0) = goal_position;
      output_(1) = goal_velocity;
    }
  }

  return output_;
}

void AsymmetricTrapezoidProfile::CalculateTimes(
    double distance_to_target, double goal_velocity,
    Eigen::Matrix<double, 2, 1> current) {
  if (distance_to_target == 0) {
    // We're there. Stop everything.
    // TODO(aschuh): Deal with velocity not right.
    acceleration_time_ = 0;
    acceleration_ = 0;
    constant_time_ = 0;
    deceleration_time_ = 0;
    deceleration_ = 0;
    deceleration_reversal_time_ = 0;
    deceleration_reversal_ = 0;
    return;
  } else if (distance_to_target < 0) {
    // Recurse with everything inverted.
    current(1) *= -1;
    CalculateTimes(-distance_to_target, -goal_velocity, current);
    acceleration_ *= -1;
    deceleration_ *= -1;
    deceleration_reversal_ *= -1;
    return;
  }

  constant_time_ = 0;
  acceleration_ = maximum_acceleration_;

  // Calculate the fastest speed we could get going to by the distance to
  // target.  We will have normalized everything out to be a positive distance
  // by now so we never have to deal with going "backwards".
  double maximum_acceleration_velocity =
      distance_to_target * 2 * std::abs(acceleration_) +
      current(1) * current(1);
  CHECK_GE(maximum_acceleration_velocity, 0);
  maximum_acceleration_velocity = sqrt(maximum_acceleration_velocity);

  // If we could get going faster than the target, we will need to decelerate
  // after accelerating.
  if (maximum_acceleration_velocity > goal_velocity) {
    deceleration_ = -maximum_deceleration_;
  } else {
    // We couldn't get up to speed by the destination.  Set our decel to
    // accelerate to keep accelerating to get up to speed.
    //
    // Note: goal_velocity != 0 isn't well tested, use at your own risk.
    LOG(FATAL) << "Untested";
    deceleration_ = maximum_acceleration_;
  }

  // If we are going away from the goal, we will need to change directions.
  if (current(1) < 0) {
    deceleration_reversal_time_ = current(1) / deceleration_;
    deceleration_reversal_ = -deceleration_;
  } else if ((goal_velocity - current(1)) * (goal_velocity + current(1)) <
             2.0 * deceleration_ * distance_to_target) {
    // Then, can we stop in time if we get after it?  If so, we don't need to
    // decel first before running the profile.
    deceleration_reversal_time_ = -current(1) / deceleration_;
    deceleration_reversal_ = deceleration_;
  } else {
    // Otherwise, we are all good and don't need to handle reversing.
    deceleration_reversal_time_ = 0.0;
    deceleration_reversal_ = 0.0;
  }

  current(0) += current(1) * deceleration_reversal_time_ +
                0.5 * deceleration_reversal_ * deceleration_reversal_time_ *
                    deceleration_reversal_time_;
  current(1) += deceleration_reversal_ * deceleration_reversal_time_;
  // OK, now we've compensated for slowing down.

  // We now know the top velocity we can get to.
  const double top_velocity = sqrt(
      (distance_to_target + (current(1) * current(1)) / (2.0 * acceleration_) +
       (goal_velocity * goal_velocity) / (2.0 * deceleration_)) /
      (-1.0 / (2.0 * deceleration_) + 1.0 / (2.0 * acceleration_)));

  // If it can go too fast, we now know how long we get to accelerate for and
  // how long to go at constant velocity.
  if (top_velocity > maximum_velocity_) {
    acceleration_time_ =
        (maximum_velocity_ - current(1)) / maximum_acceleration_;
    constant_time_ = ((-0.5 * maximum_acceleration_ * acceleration_time_ *
                           acceleration_time_ -
                       current(1) * acceleration_time_) +
                      distance_to_target +
                      (goal_velocity * goal_velocity -
                       maximum_velocity_ * maximum_velocity_) /
                          (2.0 * maximum_deceleration_)) /
                     maximum_velocity_;
  } else {
    acceleration_time_ = (top_velocity - current(1)) / acceleration_;
  }

  CHECK_GT(top_velocity, -maximum_velocity_);

  if (current(1) > maximum_velocity_) {
    constant_time_ = 0;
    acceleration_time_ = 0;
  }

  deceleration_time_ =
      (goal_velocity - ::std::min(top_velocity, maximum_velocity_)) /
      deceleration_;
  if (acceleration_time_ <= 0) acceleration_time_ = 0;
  if (constant_time_ <= 0) constant_time_ = 0;
  if (deceleration_time_ <= 0) deceleration_time_ = 0;
}

}  // namespace aos::util
